lib/mongoose_push.ex

```diff
@@ -1,18 +1,106 @@
 defmodule MongoosePush do
   @moduledoc """
-  Documentation for MongoosePush.
+  MongoosePush is simple (seriously) service providing ability to send push
+  notification to `FCM` (Firebase Cloud Messaging) and/or
+  `APNS` (Apple Push Notification Service). What makes it cool is not only
+  simplicity but also support for newest and fastest `HTTP/2` based APIs
+  for both services.
+
+  At this moment only those two services are supported but in future
+  MongoosePush may and probably will support even more Push Notification Services.
   """
 
-  @doc """
-  Hello world.
+  require Logger
+  import MongoosePush.Application
+  alias Pigeon.GCM
+  alias Pigeon.APNS
+
+  @typedoc "Available keys in `request` map"
+  @type req_key :: :service | :body | :title | :bagde | :mode | :tag |
+                   :topic | :click_action
 
-  ## Examples
+  @typedoc "Raw push request. The keys: `:service`, `:body` and `:title` are required"
+  @type request :: %{req_key => atom | String.t | integer}
+
+  @type service :: :fcm | :apns
+  @type mode :: :dev | :prod
+
+  @doc """
+  Push notification defined by `request` to device with `device_id`.
+  `request` has to define at least `:service` type (`:fcm` or `:apns`) and
+  both message `:title` and its `:body`.
 
-      iex> MongoosePush.hello
-      :world
+  `:tag` is option
+  specific to FCM service, while `:topic` and `:bagde` are specific to APNS
+  (please consult their API for more informations).
 
+  `:mode` option is also specific to APNS but it only selects appropriate
+  worker pool (with `:mode` set to either `:prod` or `:dev`).
+  Default value to `:mode` is `:prod`.
   """
-  def hello do
-    :world
+  @spec push(String.t, request) :: :ok | {:error, term}
+  def push(device_id, %{:service => service} = request) do
+      mode = Map.get(request, :mode, :prod)
+      [pool | _] = pools_by_mode(service, mode)
```

Comment:
It looks like the functions on lines 44 and 45 are consructing a worker name. For me this level of abstraction doesn't belong to this function.